libp2p/connmanager.nim

```diff
@@ -331,35 +345,120 @@ proc storeConn*(c: ConnManager, conn: Connection) =
   ##
 
   if isNil(conn):
-    raise newException(CatchableError, "connection cannot be nil")
+    raise newException(CatchableError, "Connection cannot be nil")
 
-  if conn.closed() or conn.atEof():
-    trace "Can't store dead connection", conn
-    raise newException(CatchableError, "can't store dead connection")
+  if conn.closed or conn.atEof:
+    raise newException(CatchableError, "Connection closed or EOF")
 
   if isNil(conn.peerInfo):
-    raise newException(CatchableError, "empty peer info")
+    raise newException(CatchableError, "Empty peer info")
+
+  # Launch on close listener
+  # All the errors are handled inside `onClose()` procedure.
+  asyncSpawn c.onClose(conn)
```

Comment:
should `onClose` run even if `c.maxPeerConns` is violated? if yes, why?

Reply:
also, why should it _not_ be run for connections without `peerInfo`?

Reply:
So, we don't explicitly disconnect the peer there, we only throw an exception and allow the caller to decide what needs to happen to the peer, so from that standpoint the connection, nor the peer is immediately dropped. Maybe we should drop the peer here explicitly?

Reply:
the question here is what `onClose` does / should do, and to which connections? to those that were stored in the conn manager? to some connections randomly?

Reply:
`onClose` cleans up after the connection is closed, it works on the connection that is being passed to it. But it doesn't trigger until the connection has been added to connection manager, so it won't trigger in the case of `TooManyConnections` because it was never added to the manager in the first place.

Reply:
but that's a leak then: the task gets spawned before the peer connection check

Reply:
Ah, interesting. Seems like I changed it in this PR and this would indeed leak, good catch.